smalltalksrc/Slang/TMethod.class.st

```diff
@@ -2031,6 +2077,30 @@ TMethod >> isVolatile [
 	^ self hasProperties and: [properties includesKey: #volatile]
 ]
 
+{ #category : 'inlining support' }
+TMethod >> iterateThrought: aNode in: aCodeGen [
+	"iterate throuht an AST and supress unused variables and constants, since we don't do anything when meeting a send (except in some cases), assignment or return node the only variables and constants we come accross are alone in their lines"
+	| didSomething |
+	didSomething := false.
+	aNode isStatementList ifTrue: [
+		aNode children do: [ :subNode |
+			subNode ifNotNil: [
+				didSomething := didSomething or: [ self iterateThrought: subNode in: aCodeGen ] ] ] ].
+	aNode isSwitch ifTrue: [ 
+		aNode cases do: [ :case |
+			didSomething := didSomething or: [self iterateThrought: case second in: aCodeGen] ] ].
+	(self isStructAccessorWithNoSideEffect: aNode in: aCodeGen) ifTrue: [
+		self cleanASTOf: aNode.
+		^ true ].
+	(self isMaybeSendWithoutSideEffect: aNode) ifTrue: [
+		aNode arguments do: [ :subNode |
+			didSomething := didSomething or: [self iterateThrought: subNode in: aCodeGen ] ]].
+	(aNode isVariable or: aNode isConstant) ifTrue: [
+		self cleanASTOf: aNode.
+		^ true ].
+	^ didSomething
```

Comment:
Can we exploit better the polymorphism in here?

Reply:
Either refactor this as `removeDeadCodeUsing:` or introduce it in a visitor pattern